# Analisis Hasil Pencarian Taylor & Francis (Literature Review)

Notebook ini membaca dan mengeksplorasi dataset hasil pencarian artikel ilmiah dari **Taylor & Francis** (`Taylor and Francis search results (12 September 2026).csv`).

### Alur Kerja:
1. **Import Library** — Memuat pustaka yang dibutuhkan (`pandas`, `os`, dll.)
2. **Load Dataset** — Membaca file CSV secara dinamis
3. **Pemeriksaan Struktur Data** — Mengecek tipe data, missing values, dan dimensi dataset
4. **Pratinjau Data** — Menampilkan sampel baris data
5. **Eksplorasi Statistik Sederhana** — Distribusi tahun publikasi, jurnal teratas, dan artikel paling banyak disitasi

## Langkah 1: Import Library

Memuat pustaka `os` untuk manajemen path dan `pandas` untuk manipulasi data tabular.

In [4]:
import os
import pandas as pd
import numpy as np

print("Library berhasil diimpor.")

Library berhasil diimpor.


## Langkah 2: Menentukan Path & Membaca Dataset

Mendefinisikan lokasi file `Taylor and Francis search results (12 September 2026).csv` secara dinamis agar kompatibel saat dijalankan dari berbagai direktori kerja.

In [5]:
# Pencarian path file secara fleksibel
possible_paths = [
    'Taylor and Francis search results (12 September 2026).csv',
    os.path.join('Crawling', 'Taylor and Francis search results (12 September 2026).csv'),
    os.path.join(os.path.dirname(os.path.abspath('__file__')), 'Taylor and Francis search results (12 September 2026).csv')
]

csv_path = None
for p in possible_paths:
    if os.path.exists(p):
        csv_path = p
        break

if csv_path is None:
    raise FileNotFoundError("File CSV tidak ditemukan!")

print(f"Memuat data dari: {csv_path}")
df = pd.read_csv(csv_path)
print(f"Data berhasil dimuat! Dimensi: {df.shape[0]} baris, {df.shape[1]} kolom")

Memuat data dari: Taylor and Francis search results (12 September 2026).csv
Data berhasil dimuat! Dimensi: 691 baris, 12 kolom


## Langkah 3: Pemeriksaan Struktur Data & Missing Values

Mengevaluasi kolom-kolom yang tersedia, tipe data, serta persentase nilai kosong (*missing values*) pada masing-masing fitur.

In [6]:
# Ringkasan kolom, tipe data, dan missing values
summary_df = pd.DataFrame({
    'Tipe Data': df.dtypes,
    'Nilai Terisi (Non-Null)': df.notnull().sum(),
    'Nilai Kosong (NaN)': df.isnull().sum(),
    'Persentase Kosong (%)': (df.isnull().sum() / len(df) * 100).round(2)
})
summary_df

,Tipe Data,Nilai Terisi (Non-Null),Nilai Kosong (NaN),Persentase Kosong (%)
Article title,str,691,0,0.00
Authors,str,677,14,2.03
Journal title,str,691,0,0.00
Volume,float64,606,85,12.30
Issue,str,605,86,12.45
Pages,str,637,54,7.81
Volume year,int64,691,0,0.00
DOI,str,691,0,0.00
Published online date,str,691,0,0.00
URL,str,691,0,0.00


## Langkah 4: Pratinjau Data (Head)

Menampilkan 5 baris pertama dari dataset untuk melihat isi kolom seperti judul artikel, penulis, nama jurnal, DOI, jumlah views, dan sitasi.

In [7]:
# Pratinjau 5 data pertama
df.head()

,Article title,Authors,Journal title,Volume,Issue,Pages,Volume year,DOI,Published online date,URL,Article views (as of 12-Sep-2026),Crossref citations (as of 12-Sep-2026)
0,Affective design using machine learning: a sur...,"Bilal Abu-Salih, Kit Yan Chan, Chris K Y Fung,...",International Journal of Computer Integrated M...,33.0,7,p645 - p669,2020,10.1080/0951192x.2018.1526412,04-Oct-2018,https://doi.org/10.1080/0951192x.2018.1526412,1537.0,53.0
1,Big data analytics and big data science: a survey,"Hong Chen, Yong Chen, Anjee Gorkhali, Ling Li,...",Journal of Management Analytics,3.0,1,p1 - p42,2016,10.1080/23270012.2016.1141332,26-Feb-2016,https://doi.org/10.1080/23270012.2016.1141332,3910.0,83.0
2,Knowledge generation from Big Data for new pro...,"Gioconda Mele, Giuseppina Passiante, Donata Se...",Knowledge Management Research & Practice,21.0,4,p892 - p907,2023,10.1080/14778238.2022.2094292,28-Jul-2022,https://doi.org/10.1080/14778238.2022.2094292,555.0,5.0
3,BIG data - BIG gains? Understanding the link b...,"Thomas Niebel, Fabienne Rasel, Steffen Viete",Economics of Innovation and New Technology,28.0,3,p296 - p316,2019,10.1080/10438599.2018.1493075,17-Jul-2018,https://doi.org/10.1080/10438599.2018.1493075,32882.0,158.0
4,Big data and the transformation of operations ...,"G Graham, F Li, A Nucciarelli, S Roden",Production Planning & Control,28.0,11-12,p929 - p944,2017,10.1080/09537287.2017.1336792,11-Jul-2017,https://doi.org/10.1080/09537287.2017.1336792,3172.0,84.0


## Langkah 5: Eksplorasi Singkat

Melihat distribusi tahun publikasi, jurnal teratas, serta artikel dengan sitasi Crossref tertinggi.

In [8]:
print("=== 5 Jurnal dengan Artikel Terbanyak ===")
print(df['Journal title'].value_counts().head(5))

print("\n=== Distribusi Tahun Publikasi ===")
print(df['Volume year'].value_counts().sort_index(ascending=False).head(10))

print("\n=== Top 5 Artikel Paling Banyak Disitasi ===")
top_cited = df.nlargest(5, 'Crossref citations (as of 12-Sep-2026)')[['Article title', 'Journal title', 'Volume year', 'Crossref citations (as of 12-Sep-2026)']]
top_cited

=== 5 Jurnal dengan Artikel Terbanyak ===
Journal title
International Journal of Computer Integrated Manufacturing    116
Journal of the Operational Research Society                    51
International Journal of Production Research                   36
Knowledge Management Research & Practice                       32
European Journal of Information Systems                        29
Name: count, dtype: int64

=== Distribusi Tahun Publikasi ===
Volume year
5187     85
2026     87
2025    102
2024     82
2023     68
2022     75
2021     34
2020     44
2019     38
2018     41
Name: count, dtype: int64

=== Top 5 Artikel Paling Banyak Disitasi ===


,Article title,Journal title,Volume year,Crossref citations (as of 12-Sep-2026)
622,Industry 4.0: state of the art and future trends,International Journal of Production Research,2018,2641.0
523,Artificial Intelligence and the Public Sector-...,International Journal of Public Administration,2019,1033.0
11,Creating Strategic Business Value from Big Dat...,Journal of Management Information Systems,2018,760.0
531,Sustainable manufacturing in Industry 4.0: an ...,International Journal of Production Research,2020,710.0
311,"Algorithmic bias: review, synthesis, and futur...",European Journal of Information Systems,2022,641.0


---

## Langkah 6: Web Crawling Abstrak via DOI

Mengambil teks abstrak untuk setiap artikel menggunakan DOI yang tersedia di dataset.

### Strategi Pengambilan Abstrak (Multi-Source Fallback):

| Prioritas | Sumber | Keterangan |
|-----------|--------|------------|
| 1 | **Semantic Scholar API** | Gratis, no auth, ~70% coverage untuk Taylor & Francis |
| 2 | **Crossref API** | Fallback; menyediakan abstrak pada beberapa publisher |
| 3 | `NOT_FOUND` | Artikel tidak ditemukan di kedua sumber |

> **Catatan**: Taylor & Francis (tandfonline.com) memblokir request langsung (HTTP 403), sehingga pendekatan API digunakan sebagai alternatif yang lebih andal dan legal.

### Langkah 6.1: Import Library Tambahan

In [9]:
import requests
import time
import re
import json
from pathlib import Path

print("Library crawling berhasil diimpor.")

Library crawling berhasil diimpor.


### Langkah 6.2: Fungsi Pengambilan Abstrak & Pembersihan Teks

Fungsi utama:
- **`clean_abstract(text)`** — Membersihkan kata awalan "ABSTRACT" / "Abstract" (dan tanda baca pemisah) pada kata pertama teks abstrak
- **`fetch_from_semantic_scholar(doi)`** — Mengambil abstrak dari [Semantic Scholar](https://api.semanticscholar.org/)
- **`fetch_from_crossref(doi)`** — Fallback menggunakan [Crossref API](https://api.crossref.org/)
- **`get_abstract(doi)`** — Orkestrator: mencoba Semantic Scholar dahulu, fallback ke Crossref, serta otomatis membersihkan teks abstrak

In [16]:
SEMANTIC_SCHOLAR_BASE = "https://api.semanticscholar.org/graph/v1/paper/DOI:{doi}?fields=title,abstract"
CROSSREF_BASE         = "https://api.crossref.org/works/{doi}"

HEADERS_CROSSREF = {
    "User-Agent": "AbstractCrawler/1.0 (mailto:your.email@example.com)",
    "Accept": "application/json",
}


def clean_abstract(text: str) -> str | None:
    """
    Membersihkan kata awalan 'ABSTRACT' atau 'Abstract' (beserta tanda baca pemisah seperti :, -, .)
    yang sering muncul sebagai kata pertama pada teks abstrak dari publisher/API.
    """
    if not text or not isinstance(text, str):
        return text
    # Hapus kata 'abstract' / 'abstracts' di awal teks, tanda baca pemisah (: - . dsb), atau jika langsung menempel kata huruf besar
    cleaned = re.sub(r'^\s*abstracts?(?:[\s:\.\-–—]+|(?=[A-Z])|$)', '', text, flags=re.IGNORECASE)
    cleaned = cleaned.strip()
    return cleaned if cleaned else None


def strip_jats_xml(text: str) -> str:
    """Bersihkan tag JATS XML yang kadang muncul di abstrak Crossref."""
    if not text:
        return text
    cleaned = re.sub(r'<[^>]+>', ' ', text)
    return re.sub(r'\s+', ' ', cleaned).strip()


def fetch_from_semantic_scholar(doi: str, timeout: int = 15) -> str | None:
    """Ambil abstrak dari Semantic Scholar API menggunakan DOI."""
    try:
        url = SEMANTIC_SCHOLAR_BASE.format(doi=doi)
        resp = requests.get(url, timeout=timeout)
        if resp.status_code == 200:
            data = resp.json()
            abstract = data.get("abstract")
            if abstract and len(abstract.strip()) > 20:
                return abstract.strip()
        elif resp.status_code == 429:
            print(f"    [SemanticScholar] Rate limit hit on {doi}, waiting 10s...")
            time.sleep(10)
    except Exception as e:
        print(f"    [SemanticScholar] Error for {doi}: {e}")
    return None


def fetch_from_crossref(doi: str, timeout: int = 15) -> str | None:
    """Ambil abstrak dari Crossref API sebagai fallback."""
    try:
        url = CROSSREF_BASE.format(doi=doi)
        resp = requests.get(url, headers=HEADERS_CROSSREF, timeout=timeout)
        if resp.status_code == 200:
            work = resp.json().get("message", {})
            abstract = work.get("abstract")
            if abstract and len(abstract.strip()) > 20:
                return strip_jats_xml(abstract)
    except Exception as e:
        print(f"    [Crossref] Error for {doi}: {e}")
    return None


def get_abstract(doi: str) -> dict:
    """
    Orkestrator multi-source untuk mengambil abstrak.
    Urutan: Semantic Scholar → Crossref → NOT_FOUND

    Returns:
        dict dengan keys: 'abstract', 'source', 'status'
    """
    abstract = fetch_from_semantic_scholar(doi)
    if abstract:
        return {"abstract": clean_abstract(abstract), "source": "SemanticScholar", "status": "FOUND"}

    abstract = fetch_from_crossref(doi)
    if abstract:
        return {"abstract": clean_abstract(abstract), "source": "Crossref", "status": "FOUND"}

    return {"abstract": None, "source": None, "status": "NOT_FOUND"}


print("Fungsi pengambilan abstrak dan pembersihan teks berhasil didefinisikan.")

Fungsi pengambilan abstrak dan pembersihan teks berhasil didefinisikan.


### Langkah 6.3: Uji Coba (Test Run) pada 5 Artikel Pertama

Verifikasi bahwa fungsi crawling bekerja dengan benar sebelum menjalankan ke seluruh dataset.

In [11]:
test_dois = df['DOI'].dropna().head(5).tolist()

print("=== TEST RUN: 5 Artikel Pertama ===")
for i, doi in enumerate(test_dois, 1):
    result = get_abstract(doi)
    title = df.loc[df['DOI'] == doi, 'Article title'].values[0]
    print(f"\n[{i}] {title[:70]}...")
    print(f"    DOI    : {doi}")
    print(f"    Status : {result['status']} | Source: {result['source']}")
    if result['abstract']:
        print(f"    Abstrak: {result['abstract'][:150]}...")
    else:
        print(f"    Abstrak: (tidak ditemukan)")
    time.sleep(0.5)

=== TEST RUN: 5 Artikel Pertama ===

[1] Affective design using machine learning: a survey and its prospect of ...
    DOI    : 10.1080/0951192x.2018.1526412
    Status : FOUND | Source: SemanticScholar
    Abstrak: ABSTRACT Customer satisfaction in purchasing new products is an important issue that needs to be addressed in today’s competitive markets. A product w...

[2] Big data analytics and big data science: a survey...
    DOI    : 10.1080/23270012.2016.1141332
    Status : NOT_FOUND | Source: None
    Abstrak: (tidak ditemukan)

[3] Knowledge generation from Big Data for new product development: a stru...
    DOI    : 10.1080/14778238.2022.2094292
    Status : FOUND | Source: SemanticScholar
    Abstrak: ABSTRACT This study aims to provide a contribution to the systematisation of the state-of-the art of the literature on Big Data for the process of New...

[4] BIG data - BIG gains? Understanding the link between big data analytic...
    DOI    : 10.1080/10438599.2018.1493075
    

### Langkah 6.4: Crawling Seluruh Dataset

Menjalankan pengambilan abstrak untuk semua **691 artikel** dengan:
- Delay antar request: **0.35 detik** (menghindari rate-limit)
- Progress log setiap 50 artikel
- Auto-resume: jika file hasil sudah ada, artikel yang sudah diproses dilewati
- Simpan checkpoint sementara setiap 50 artikel

In [12]:
DELAY_BETWEEN_REQUESTS = 0.35
CHECKPOINT_EVERY = 50
OUTPUT_PATH = Path('crawling_results.json')

def load_checkpoint(path: Path) -> dict:
    """Load hasil crawling yang sudah ada (untuk resume)."""
    if path.exists():
        with open(path, 'r', encoding='utf-8') as f:
            return json.load(f)
    return {}


def save_checkpoint(results: dict, path: Path):
    """Simpan checkpoint crawling ke JSON."""
    with open(path, 'w', encoding='utf-8') as f:
        json.dump(results, f, ensure_ascii=False, indent=2)


all_dois = df['DOI'].dropna().tolist()
results = load_checkpoint(OUTPUT_PATH)

already_done = len(results)
to_process   = [d for d in all_dois if d not in results]

print(f"Total DOI     : {len(all_dois)}")
print(f"Sudah selesai : {already_done}")
print(f"Perlu diproses: {len(to_process)}")
print()

for i, doi in enumerate(to_process, 1):
    result = get_abstract(doi)
    results[doi] = result

    if i % CHECKPOINT_EVERY == 0 or i == len(to_process):
        save_checkpoint(results, OUTPUT_PATH)
        found_count = sum(1 for v in results.values() if v['status'] == 'FOUND')
        print(f"[Progress {i}/{len(to_process)}] Tersimpan. Found: {found_count}/{len(results)} ({found_count/len(results)*100:.1f}%)")

    time.sleep(DELAY_BETWEEN_REQUESTS)

save_checkpoint(results, OUTPUT_PATH)
print(f"\nCrawling selesai! Total hasil disimpan ke: {OUTPUT_PATH.resolve()}")

Total DOI     : 691
Sudah selesai : 0
Perlu diproses: 691

[Progress 50/691] Tersimpan. Found: 38/50 (76.0%)
[Progress 100/691] Tersimpan. Found: 73/100 (73.0%)
[Progress 150/691] Tersimpan. Found: 110/150 (73.3%)
[Progress 200/691] Tersimpan. Found: 151/200 (75.5%)
    [SemanticScholar] Rate limit hit on 10.1080/09654313.2022.2135368, waiting 10s...
    [SemanticScholar] Rate limit hit on 10.1080/0144929x.2019.1641663, waiting 10s...
    [SemanticScholar] Rate limit hit on 10.1080/13683500.2025.2581155, waiting 10s...
    [SemanticScholar] Rate limit hit on 10.1080/13682199.2018.1532670, waiting 10s...
    [SemanticScholar] Rate limit hit on 10.1080/0951192x.2017.1407447, waiting 10s...
[Progress 250/691] Tersimpan. Found: 180/250 (72.0%)
[Progress 300/691] Tersimpan. Found: 215/300 (71.7%)
[Progress 350/691] Tersimpan. Found: 254/350 (72.6%)
[Progress 400/691] Tersimpan. Found: 300/400 (75.0%)
    [SemanticScholar] Rate limit hit on 10.1080/23311975.2025.2471538, waiting 10s...
    [

### Langkah 6.5: Integrasi Hasil ke DataFrame, Pembersihan Kolom Abstrak & Ekspor CSV

Menggabungkan hasil crawling ke dalam DataFrame utama, memastikan kata awalan `abstract` pada kolom `abstract` dibersihkan, dan mengekspor ke file CSV baru.

In [17]:
df_result = df.copy()

# Pemetaan hasil crawling
df_result['abstract']        = df_result['DOI'].map(lambda doi: results.get(doi, {}).get('abstract'))
df_result['abstract_source'] = df_result['DOI'].map(lambda doi: results.get(doi, {}).get('source'))
df_result['abstract_status'] = df_result['DOI'].map(lambda doi: results.get(doi, {}).get('status', 'NOT_PROCESSED'))

# Bersihkan kata 'abstract' jika muncul sebagai kata pertama di kolom abstract
df_result['abstract'] = df_result['abstract'].apply(clean_abstract)

print("=== Ringkasan Hasil Crawling ===")
print(df_result['abstract_status'].value_counts())
print()
print("=== Distribusi Sumber Abstrak ===")
print(df_result['abstract_source'].value_counts(dropna=False))
print()

coverage = (df_result['abstract_status'] == 'FOUND').sum() / len(df_result) * 100
print(f"Coverage Abstrak: {coverage:.1f}%")

OUTPUT_CSV = 'taylor_francis_with_abstracts.csv'
df_result.to_csv(OUTPUT_CSV, index=False, encoding='utf-8-sig')
print(f"\nDataset dengan abstrak bersih disimpan ke: {OUTPUT_CSV}")

=== Ringkasan Hasil Crawling ===
abstract_status
FOUND        503
NOT_FOUND    188
Name: count, dtype: int64

=== Distribusi Sumber Abstrak ===
abstract_source
SemanticScholar    503
NaN                188
Name: count, dtype: int64

Coverage Abstrak: 72.8%

Dataset dengan abstrak bersih disimpan ke: taylor_francis_with_abstracts.csv


### Langkah 6.6: Pratinjau Hasil Akhir

Menampilkan sampel artikel dengan abstrak yang berhasil diperoleh.

In [18]:
df_found = df_result[df_result['abstract_status'] == 'FOUND'][['Article title', 'Journal title', 'Volume year', 'abstract', 'abstract_source']]

print(f"Artikel dengan abstrak: {len(df_found)} dari {len(df_result)}")
df_found.head(5)

Artikel dengan abstrak: 503 dari 691


,Article title,Journal title,Volume year,abstract,abstract_source
0,Affective design using machine learning: a sur...,International Journal of Computer Integrated M...,2020,Customer satisfaction in purchasing new produc...,SemanticScholar
2,Knowledge generation from Big Data for new pro...,Knowledge Management Research & Practice,2023,This study aims to provide a contribution to t...,SemanticScholar
3,BIG data - BIG gains? Understanding the link b...,Economics of Innovation and New Technology,2019,This paper analyzes the relationship between f...,SemanticScholar
4,Big data and the transformation of operations ...,Production Planning & Control,2017,Big Data has been hailed as the ‘next big thin...,SemanticScholar
5,A survey of machine learning in additive manuf...,International Journal of Computer Integrated M...,2023,"Thirty years into its development, additive ma...",SemanticScholar
